### DISCUSSION TITLE
X11 Guide

### DISCUSSION THREAD

**Discussion by rusty-snake** [netblue30/firejail#4451] [Wiki]
Discussion for discussions about https://github.com/netblue30/firejail/wiki/X11-Guide.

Predecessor: #3602

**Comment by crocket** [comment]
`--x11=xpra` doesn't really launch a window, and it is very slow to set up an xpra sandbox for some reason.

Xephyr is said to have performance issue.

Does x11 security extension work?

**Reply by crocket** [reply]
I made xpra work by adding my user to xpra group. But, alacritty is very slow on xpra. I don't know why it is slow.
I also hate seeing xpra splash screen that I see with `--x11=xpra`.

**Reply by rusty-snake** [reply]
#640

**Reply by crocket** [reply]
It seems I can't use xpra... Does xephyr have performance issues?

**Reply by rusty-snake** [reply]
There is #3702 but IDK if this is a config problem or a general problem.

**Reply by crocket** [reply]
I don't use virtualgl.

**Reply by crocket** [reply]
It seems xephyr doesn't give me a resizable window. A tiling window manager like sway doesn't work well with xephyr.
It seems I'm stuck with either XWayland or xpra.
How can I make xpra fast? Xpra is reasonably fast outside firejail. But, it is very slow inside firejail. Why?

**Reply by crocket** [reply]
Anyway, it seems sandboxing X11 involves too much trouble. I will just migrate as many applications as possible over to wayland.

**Comment by matu3ba** [comment]
Author of x11 guide (that is based on test experience of the netblue30 article) here.

"Xephyr is said to have performance issue." 
Not true, I wrote after testing "occasionally has lags, where you cant move your mouse for a while." So its usable, but the mouse may stutter sometimes.
Do you have suggestions how to make this more clear in the wiki?

**Reply by crocket** [reply]
Is xephyr a lot faster than xpra inside a firejail sandbox?

> I wrote after testing "occasionally has lags, where you cant move your mouse for a while." So its usable, but the mouse may stutter sometimes. Do you have suggestions how to make this more clear in the wiki?

There is no performance issue apart from the issue that mouse cursor occasionally stutters.

**Reply by matu3ba** [reply]
done.

**Comment by totaam** [comment]
> I made xpra work by adding my user to xpra group

That's only needed if `$XDG_RUNTIME_DIR` is not created.
And even then, xpra should gracefully fallback to something like `/tmp`.

> I also hate seeing xpra splash screen that I see with --x11=xpra.

`--splash=no`

> How can I make xpra fast? Xpra is reasonably fast outside firejail.

When running locally, it should be almost indistinguishable from native.
(bar startup cost - which can be drastically reduced via configuration: no splash, no opengl, fixed fakexinerama path, etc)

> But, it is very slow inside firejail. Why?

My guess is that mmap is disabled so all the screen updates end up being encoded and forwarded over a unix domain socket. Even a virtual machine would be faster than this!

**Reply by rusty-snake** [reply]
xpra.profile does not block `mmap` via seccomp.

**Reply by totaam** [reply]
> xpra.profile does not block `mmap` via seccomp.

My bad. Just tried it and mmap is enabled, performance is pretty good here on Fedora 34 with Firefox running at 4K. Only the occasional framedrop watching youtube fullscreen.

**Reply by crocket** [reply]
$XDG_RUNTIME_DIR exists.

How can I pass `--splash=no` to xpra through firejail? `firejail --noprofile --x11=xpra alacritty`

Can you give me command examples?

**Reply by rusty-snake** [reply]
https://github.com/netblue30/firejail/blob/7aec067d5cc7564ad54d3c520ae97c474d628937/etc/firejail.config#L151-L152

**Comment by totaam** [comment]
> How can I pass --splash=no to xpra through firejail?

I don't know how to do that with firejail, but you can do that easily enough with xpra:
```
echo "splash = no" >> ~/.config/xpra/xpra.conf
```
(older versions used `.xpra/xpra.conf`)

**Reply by crocket** [reply]
I disabled xpra splash. But, `--x11=xpra` slows down startup and rendering of alacritty significantly to the point where I think a VTE-based terminal emulator on wayland backend is going to be much better.

**Reply by totaam** [reply]
> slows down startup

Default config uses Xdummy, Xvfb is faster.
But in any case, it will be slower than native: around 2 second is standard.

> I think a VTE-based terminal emulator

Yes, why go through all this trouble for a terminal emulator?

**Reply by crocket** [reply]
X11 is evil. After replacing alacritty with a VTE-based terminal emulator and sxiv with imv, emacs is the last remaining X11 program on my system. I can accept one X11 program on sway.

Even emacs is going to be a native citizen on wayland soon.

**Comment by jjakob** [comment]
I managed to get xclipsync to work with Xephyr. A modification to the xclipsync script was needed in my case to use the exact binary name wish8.6 instead of just wish, which is a symlink. Maybe because I used private-bin or because I blacklisted /usr/bin in my profile, /usr/bin was "permission denied".

1. download xclipsync and xclipfrom and put them in /usr/local/bin, chmod +x (they're <40 LOC each)
2. edit xclipsync, change lines 37 and 38: replace "wish" with "wish8.6" or the name of the wish binary (find it with `command -v wish`) - this is needed probably because I blacklisted /usr/bin in the profile which meant all binaries were only accessible in /bin and /sbin, the wish symlink pointed to the blacklisted /usr/bin/wish8.6
3. if you use private-bin you need to add: `xclipsync,xclipfrom,xclip,dirname,wish*`
4. if you use private-lib you need to add: `tcl*,tk*`
5. change your command line to add `xclipsync &` before starting the WM or browser

With Xephyr you usually also need a WM, I needed it because the browser window did not resize itself to the resolution of the Xephyr display but stayed smaller. I used dwm because that's what I already had installed but is overkill. Other possible choices of minimal WMs might be matchbox, ratpoison, 2bwm. You just want something that handles window resizing, no other features. kiosk-wm might also work, it's even smaller than the rest.

**Reply by matu3ba** [reply]
Feel free to edit the wiki page with instruction steps of the X11 Guide at "TODO: [xclipsync](https://github.com/apenwarr/xclipsync) appears not to work or write cli for clipboard handling?".

I'll be switching to Arcan and do not plan to complete the guide to work around these problems.
